[FLINK-19462][checkpointing] Update failed checkpoint stats

rkhachatryan opened the pull request:
## What is the purpose of the change

Update checkpoint statistics (shown in the web UI) even after a checkpoint fails
(this would facilitate investigation of issues with slow checkpointing).

With this change, failed checkpoint stats is updated when:
1. Subtask acks a checkpoint too late or after some other failure. `AsyncCheckpointRunnable` completes normally and reports snapshot as usual. `CheckpointCoordinator` was updated to handle these calls
1. Subtask receives abortion notification and cancels the runnable before it completes. In this case it only reports the metrics. Both TM and JM sides were updated and a **new RPC added**

## Verifying this change

This change added tests and can be verified as follows:
- `CheckpointCoordinatorTest.testCheckpointStatsUpdatedAfterFailure`
- `CheckpointCoordinatorTest.testAbortedCheckpointStatsUpdatedAfterFailure`
- Manually verified the change by running `DataStreamAllroundTestProgram` on local cluser:
```
execution.checkpointing.interval: 10s
execution.checkpointing.min-pause: 1s
execution.checkpointing.timeout: 1s
execution.checkpointing.tolerable-failed-checkpoints: 1000000
execution.checkpointing.unaligned: true
taskmanager.numberOfTaskSlots: 8
web.checkpoints.history: 100
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes 
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FailedCheckpointStats.java:
```diff
@@ -87,15 +72,20 @@
+                totalSubtaskCount,
+                numAcknowledgedSubtasks,
+                taskStats,
+                PendingCheckpointStatsCallback.noOp(),
```
I don't understand this. How is this supposed to be working? How stats are being reported for failed checkpoints? How are we reporting I get, that in this commit you would like to have temporary `noOp` being used, but I don't see this being replaced later? Is something missing, or is this a dead code, or am I missing something?

rkhachatryan replied:
First, `PendingCheckpointStats` is created. It has "normal" callbacks and subtasks reports update it as usual.
Upon failure it is converted to a `FailedCheckpointStats` using one of these "normal" callbacks.
This new object won't be converted so it has `noOp` callbacks. It still can be updated from subtasks.

I tried to refactor this code to have just one `CheckpointStats` class without conversions and callbacks.
But the change was to big compared to the functionality added so I dropped it.

rkhachatryan replied:
I replaced `noOp` with a static one throwing `UnsupportedOperationException`.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/SubtaskStateStats.java:
```diff
@@ -75,7 +75,6 @@
 
         checkArgument(subtaskIndex >= 0, "Negative subtask index");
         this.subtaskIndex = subtaskIndex;
-        checkArgument(stateSize >= 0, "Negative state size");
```
Why was this removed?

rkhachatryan replied:
I removed it for debugging, will restore. Thanks for pointing out!

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -1051,6 +1051,34 @@ public void acknowledgeCheckpoint(
+    @Override
+    public void reportCheckpointMetrics(
+            JobID jobID, ExecutionAttemptID attemptId, long id, CheckpointMetrics metrics) {
+        mainThreadExecutor.assertRunningInMainThread();
+
+        final CheckpointCoordinator checkpointCoordinator =
+                executionGraph.getCheckpointCoordinator();
+
+        if (checkpointCoordinator != null) {
+            ioExecutor.execute(
+                    () -> {
+                        try {
+                            checkpointCoordinator.reportStats(id, attemptId, metrics);
+                        } catch (Throwable t) {
+                            log.warn("Error while processing report checkpoint stats message", t);
+                        }
+                    });
+        } else {
+            String errorMessage =
+                    "Received ReportCheckpointStats message for job {} with no CheckpointCoordinator";
+            if (executionGraph.getState() == JobStatus.RUNNING) {
+                log.error(errorMessage, jobGraph.getJobID());
+            } else {
+                log.debug(errorMessage, jobGraph.getJobID());
+            }
+        }
+    }
```
Could we deduplicate those lines (error handling and logging) with `acknowledgeCheckpoint`? They seem to differ only by a lambda function and string messages, which maybe could be easily extracted as parameters?

rkhachatryan replied:
Good idea!
(I'll extract a  method in a separate commit as there are at least 2 existing methods already)

pnowojski reviewed (commented):
Thanks for the change @rkhachatryan . I've left a couple of comments and questions (I might haven't fully understood some things)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointMetrics.java:
```diff
@@ -46,8 +48,11 @@
+    @VisibleForTesting
     public CheckpointMetrics() {
-        this(-1L, -1L, -1L, -1L, -1L, -1L, false);
+        this(-1L, -1L, -1L, -1L, -1L, -1L, false, 0L);
```
Negative state is not allowed (and I think this is a valid check).
So some tests failed with `-1L` and I changed it to `0L`.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java:
```diff
@@ -70,9 +70,11 @@ public static void bestEffortDiscardAllStateObjects(
-    public static void discardStateFuture(Future<? extends StateObject> stateFuture)
+    public static long discardStateFuture(Future<? extends StateObject> stateFuture)
             throws Exception {
+        long stateSize = 0;
```
It would be problematic to combine `-1` with valid values.
I think we should just mention in the docs that this number is best effort if checkpoint isn't completed.

rkhachatryan commented:
Thanks for reviewing, @pnowojski.
I've addressed your feedback, PTAL.

pnowojski reviewed (changes requested):
I've tried out those changes and I think they are not working well as it is now. I've modified `WordCount` example to randomly timeout async checkpoint phase and I wanted to see what does it look like.

Two screenshots: https://imgur.com/a/sv6w2IJ

With this PR there is no way to know which task has failed checkpoint. Also async phase timed out (5 seconds) and the end to end duration is in milliseconds range for all tasks/subtasks? Async duration is also in [0ms, 5ms] range for all tasks, even the one that has timed out.

I think it's strictly necessary to:
- clearly mark which checkpoint for which subtask has failed 
- if we were not able to collect/calculate a metric, it must be `N/A` - not just `0ms`
I think it's almost must have to:
- correctly calculate the durations (end to end, sync, async, etc...) also for failed checkpoints, not just `N/A`

Otherwise this is very misleading :(

(I haven't tried it with different kind of failures, but this should be tried out before merging)

rkhachatryan commented:
Thanks a lot for trying it out.
> I think it's strictly necessary to:
> clearly mark which checkpoint for which subtask has failed

It is not always the task that fails a checkpoint. Timeout decision is made by the `CheckpointCoordinator`.
Multiple tasks can fail independently as well.
I agree that marking "failed" tasks would be useful but I don't think it's directly related to this feature or at least this PR.

> if we were not able to collect/calculate a metric, it must be N/A - not just 0ms

I don't see `0ms` on your screenshots nor while running locally. Do you mean `0 B` per operator? 
If so, why is it incorrect? (I do see non-zero size running cluster).

> correctly calculate the durations (end to end, sync, async, etc...) also for failed checkpoints, not just N/A

A checkpoint can be cancelled before even being started on some subtasks.

rkhachatryan commented:
I've updated the PR (adding 4 new commits):
1. Tasks reporting upon abort RPC are marked as `aborted` in e2e duration column
2. Only tasks that actually ACKed checkpoint are counted for ackCount and lastAckTime
3. `-1B` is shown as `-` (the same way as durations)
4. Fix the docs

![image](https://user-images.githubusercontent.com/3939322/105499876-669e6700-5cc2-11eb-8d99-b301a83a548c.png)

cc: @NicoK

NicoK commented:
So, as soon as we are through the sync phase, we will get stats (if the CP is aborted during the sync phase, that won't interrupt the sync part anyway and will wait for it to complete). If we didn't reach the sync phase yet, the timeout could be because of slowly moving barriers (no barrier was received yet) or slow alignment (some barriers received but not all). These could be derived from looking at backpressure or data skew or starting times of other subtasks or timings from previous subtasks.

I think, the current state is a good step forward and the stats look good :+1:

pnowojski commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/subtask/job-checkpoints-subtask.component.html:
```diff
@@ -99,7 +99,7 @@
         <ng-container *ngIf="subTask['status'] == 'completed'">
           <td >{{ subTask['ack_timestamp'] | date:'yyyy-MM-dd HH:mm:ss' }}</td>
-          <td>{{ subTask['end_to_end_duration'] | humanizeDuration}}</td>
+          <td>{{ subTask['end_to_end_duration'] | humanizeDuration}}<span *ngIf="subTask['aborted']">(aborted)</span></td>
```
add space in front  of `(aborted)`?

pnowojski commented on docs/ops/monitoring/checkpoint_monitoring.md:
```diff
@@ -61,7 +61,7 @@ Note that for failed checkpoints, metrics are updated on a best efforts basis an
 - **Status**: The current status of the checkpoint, which is either *In Progress* (<i aria-hidden="true" class="fa fa-circle-o-notch fa-spin fa-fw"/>), *Completed* (<i aria-hidden="true" class="fa fa-check"/>), or *Failed* (<i aria-hidden="true" class="fa fa-remove"/>). If the triggered checkpoint is a savepoint, you will see a <i aria-hidden="true" class="fa fa-floppy-o"/> symbol.
 - **Trigger Time**: The time when the checkpoint was triggered at the JobManager.
-- **Latest Acknowledgement**: The time when the latest acknowledgement for any subtask was received at the JobManager (or n/a if no acknowledgement received yet). For a failed checkpoint, this is the time from trigger timestamp to failure.
+- **Latest Acknowledgement**: The time when the latest acknowledgement for any subtask was received at the JobManager (or n/a if no acknowledgement received yet).
```
Update the `.zh.md` version as well?

rkhachatryan replied:
I don't know Chinese so I created FLINK-21122 to update it according to the [contribution guide](https://flink.apache.org/contributing/contribute-documentation.html#chinese-documentation-translation).

pnowojski reviewed (commented):
Thanks for the update @rkhachatryan. % some minor comments, there is potentially one more issue. If for example `AsynCheckpointRunnable` fails (throws an exception), I can not see any stats for any subtasks that have finished after the failure. In that case I see only `n/a` for the whole subtask, which is a bit inconsistent with timeout behaviour.

rkhachatryan commented:
Thanks for the review @pnowojski .
I've added the space and created a ticket to translate the docs.
I've also squashed the commits.

> for example AsynCheckpointRunnable fails (throws an exception), I can not see any stats for any subtasks that have finished after the failure

As discussed offline, this happens because the failed upstream doesn't sent barrier downstream.
